### PR TITLE
Added support for trim registries

### PR DESCRIPTION
**Pull request by lordtgm**
trim_material and trim_pattern registries are now fully generated from data generator and sent to the client. this should enable support for trimmed armor.

### CODE REVIEW

**Review by mworzala — COMMENTED** [review]
I am generally happy with this PR for what it is, though I think that trim materials should not work like static entries as they are here. Since they can be changed on the client it seems like it would be more fitting for them to be handled more similarly to Biomes (ie a registry where you can add more of them)

**Comment by lordtgm** [comment]
now allows dynamic creation alongside loading default ones. TrimManager::addDefaultTrimMaterials and TrimManager::addDefaultTrimPatterns can be called to include default registries.

File: src/main/java/net/minestom/server/item/armor/TrimManager.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        this.trimMaterialCache = null;
+        return this.trimMaterials.remove(trimMaterial);
+    }
+
+    public boolean addTrimPattern(TrimPattern trimPattern) {
+        this.trimMaterialCache = null;
+        return this.trimPatterns.add(trimPattern);
+    }
+
+    public boolean removePattern(TrimPattern trimPattern) {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimManager.java:90]: removeTrimPattern to match the other methods, or change them all to be addPattern, addMaterial etc

File: src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    static Collection<TrimMaterial> values() {
+        return CONTAINER.values();
+    }
+
+    public NBTCompound asNBT() {
+        return NBT.Compound(nbt -> {
+            nbt.setString("asset_name",assetName());
+            nbt.setString("ingredient",ingredient().namespace().asString());
+            nbt.setFloat("item_model_index", itemModelIndex());
+            nbt.set("override_armor_materials",NBT.Compound(overrideArmorMaterials().entrySet().stream()
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java:40]: Weird formatting

File: src/main/java/net/minestom/server/item/armor/TrimPatternImpl.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+
+record TrimPatternImpl(Registry.TrimPatternEntry registry, int id) implements TrimPattern {
+    private static final Registry.Container<TrimPattern> CONTAINER;
+    static final AtomicInteger i = new AtomicInteger();
+    static {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimPatternImpl.java:16]: formatting (add empty line or run ij formatter to do it)

File: src/main/java/net/minestom/server/item/armor/TrimPatternImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        CONTAINER = Registry.createContainer(Registry.Resource.TRIM_PATTERNS,
+                (namespace, properties) -> new TrimPatternImpl(Registry.trimPattern(namespace, properties)));
+    }
+
+    public TrimPatternImpl(Registry.TrimPatternEntry registry) {
+        this(registry, i.getAndIncrement());
+    }
+    public static TrimPattern get(String namespace) {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimPatternImpl.java:24]: formatting (add empty line or run ij formatter to do it)

File: src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+record TrimMaterialImpl(Registry.TrimMaterialEntry registry, int id) implements TrimMaterial {
+    private static final Registry.Container<TrimMaterial> CONTAINER;
+    static final AtomicInteger i = new AtomicInteger();
+    static {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java:18]: formatting (add empty line or run ij formatter to do it)

File: src/main/java/net/minestom/server/item/armor/TrimMaterial.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    default float itemModelIndex() {
+        return registry().itemModelIndex();
+    }
+
+    default @NotNull Map<String,String> overrideArmorMaterials() {
+        return registry().overrideArmorMaterials();
+    }
+    default @NotNull Component description() {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimMaterial.java:78]: formatting (add empty line or run ij formatter to do it)

File: src/main/java/net/minestom/server/item/armor/TrimManager.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                    .toList();
+
+            trimMaterialCache = NBT.Compound(Map.of(
+                    "type", NBT.String("minecraft:trim_material"),
+                    "value", NBT.List(NBTType.TAG_Compound, trimMaterials)
+            ));
+        }
+        return trimMaterialCache;
+    }
+    public NBTCompound getTrimPatternNBT() {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimManager.java:38]: formatting (add empty line or run ij formatter to do it)

File: src/main/java/net/minestom/server/item/armor/TrimManager.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class TrimManager {
+    private final Set<TrimMaterial> trimMaterials;
+    private final Set<TrimPattern> trimPatterns;
+    private NBTCompound trimMaterialCache = null;
+    private NBTCompound trimPatternCache = null;
+    public TrimManager() {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/item/armor/TrimManager.java:16]: formatting (add empty line or run ij formatter to do it)

File: code-generators/src/main/java/net/minestom/codegen/Generators.java [hunk 9]
```diff
@@ -34,6 +34,9 @@ public static void main(String[] args) {
         generator.generate(resource("sounds.json"), "net.minestom.server.sound", "SoundEvent", "SoundEventImpl", "SoundEvents");
         generator.generate(resource("custom_statistics.json"), "net.minestom.server.statistic", "StatisticType", "StatisticTypeImpl", "StatisticTypes");
         generator.generate(resource("damage_types.json"), "net.minestom.server.entity.damage", "DamageType", "DamageTypeImpl", "DamageTypes");
+        generator.generate(resource("trim_materials.json"), "net.minestom.server.item.armor", "TrimMaterial", "TrimMaterialImpl", "TrimMaterials");
+        generator.generate(resource("trim_patterns.json"), "net.minestom.server.item.armor", "TrimPattern", "TrimPatternImpl", "TrimPatterns");
+
```
Comment (mworzala) [inline code-generators/src/main/java/net/minestom/codegen/Generators.java:39]: too many spaces

**Review by mworzala — COMMENTED** [review]
Looks good besides a few minor formatting things

File: src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java (hunk 2 above)
Comment (lordtgm) [reply src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java:40]: not sure what you mean on this one. can you explain?

Comment (mworzala) [reply src/main/java/net/minestom/server/item/armor/TrimMaterialImpl.java]: just run IJ formatter (its the lack of space after the comma mostly)

**Comment by mworzala** [comment]
Looks good, thanks for doing it.

**Pull request merged**